### PR TITLE
Dwn 39926 input validation

### PR DESCRIPTION
**Pull request by hsmith-gresham**
See also https://github.com/gresham-computing/ctc/pull/1427

### CODE REVIEW

**Review by miracheta — APPROVED** [review]

**Comment by jasontse-greshamtech** [comment]
Where is this WhitelistAPI class used?

**Comment by hsmith-gresham** [comment]
> Where is this WhitelistAPI class used?

On the Auth Server Admin pages, when adding or modifying a whitelisted client

**Review by jasontse-greshamtech — APPROVED** [review]

**Review by jjuaniveson-gresham — APPROVED** [review]

File: openid-connect-server/src/main/java/org/mitre/oauth2/web/ScopeAPI.java [hunk 1]
```diff
@@ -101,7 +105,14 @@ public String updateScope(@PathVariable("id") Long id, @RequestBody String json,
 		SystemScope existing = scopeService.getById(id);
 
 		SystemScope scope = gson.fromJson(json, SystemScope.class);
-
+		try {
+			validateScope(scope);
+		} catch (ScopeException e) {
+			logger.error("updateScope failed due to ScopeException. {}", e.getMessage());
```
Comment (miracheta) [inline openid-connect-server/src/main/java/org/mitre/oauth2/web/ScopeAPI.java]: Could we add more information regarding what is that we are trying to change, id for example?

Comment (miracheta) [reply openid-connect-server/src/main/java/org/mitre/oauth2/web/ScopeAPI.java]: OK, I can see the invalid scope is part of the ScopeException

**Review by miracheta — APPROVED** [review]

File: openid-connect-server/src/main/java/org/mitre/oauth2/web/ScopeAPI.java (hunk 1 above)
Comment (hsmith-gresham) [reply openid-connect-server/src/main/java/org/mitre/oauth2/web/ScopeAPI.java]: Just confirming you're happy for me to resolve this - do you wish for any further information on these error messages?

Comment (miracheta) [reply openid-connect-server/src/main/java/org/mitre/oauth2/web/ScopeAPI.java]: I think it's OK. At least it contains the scope that failed on the exception.

File: openid-connect-server/src/main/java/org/mitre/openid/connect/web/WhitelistAPI.java [hunk 2]
```diff
@@ -167,6 +180,14 @@ public String updateWhitelistedSite(@PathVariable("id") Long id, @RequestBody St
 		}
 	}
 
+	private void validateWhitelistScopes(Set<String> scopes) throws ScopeException {
+		for (String s : scopes) {
+			if (!s.matches(characterMatcher)) {
```
Comment (jasontse-greshamtech) [inline openid-connect-server/src/main/java/org/mitre/openid/connect/web/WhitelistAPI.java]: In class ScopeAPI and here. Looks like we could be doing regex matching multiple times, maybe we can create the Pattern object ahead of time for some performance benefit (if potentially there are many scopes or the method called again).

```
private static final Pattern pattern = Pattern.compile(characterMatcher);
....
private void validateWhitelistScopes(Set<String> scopes) throws ScopeException {
    for (String s : scopes) {
        if (!pattern.matcher(s).matches()) {
            throw new ScopeException(s);
        }
    }
}
```

Comment (hsmith-gresham) [reply openid-connect-server/src/main/java/org/mitre/openid/connect/web/WhitelistAPI.java]: Good shout, adjusted

**Review by jasontse-greshamtech — APPROVED** [review]

**Review by miracheta — APPROVED** [review]

File: openid-connect-server/src/main/java/org/mitre/oauth2/web/ScopeAPI.java (hunk 1 above)
Comment (miracheta) [reply openid-connect-server/src/main/java/org/mitre/oauth2/web/ScopeAPI.java]: Sorry for being annoying, can we pass the exception and not exception.getMessage() to the logger please? I can see you are already doing that on the ClientAPI.

File: openid-connect-server/src/main/java/org/mitre/openid/connect/web/WhitelistAPI.java [hunk 3]
```diff
@@ -100,7 +105,12 @@ public String addNewWhitelistedSite(@RequestBody String jsonString, ModelMap m,
 		try {
 			json = parser.parse(jsonString).getAsJsonObject();
 			whitelist = gson.fromJson(json, WhitelistedSite.class);
-
+			validateWhitelistScopes(whitelist.getAllowedScopes());
+		} catch (ScopeException e) {
+			logger.error("addNewWhitelistedSite failed due to ScopeException. {}", e.getMessage());
```
Comment (miracheta) [inline openid-connect-server/src/main/java/org/mitre/openid/connect/web/WhitelistAPI.java]: Same here regarding e.getMessage() please

**Review by miracheta — APPROVED** [review]

File: openid-connect-server/src/main/java/org/mitre/oauth2/web/ScopeAPI.java (hunk 1 above)
Comment (hsmith-gresham) [reply openid-connect-server/src/main/java/org/mitre/oauth2/web/ScopeAPI.java]: Yep, makes sense

File: openid-connect-server/src/main/java/org/mitre/openid/connect/web/WhitelistAPI.java (hunk 3 above)
Comment (hsmith-gresham) [reply openid-connect-server/src/main/java/org/mitre/openid/connect/web/WhitelistAPI.java]: Done

**Review by miracheta — APPROVED** [review]

**Review by jasontse-greshamtech — APPROVED** [review]

**Pull request merged**